### ISSUE TITLE
When should it be possible to add previously discarded codecs in a subsequent offer?

### ISSUE THREAD

**Issue by taylor-b** [rtcweb-wg/jsep#266]
Section 5.2.2, "Subsequent Offers", says:

>    o  The m= line and corresponding "a=rtpmap" and "a=fmtp" lines MUST
>       only include codecs present in the remote description.

However, consider a situation where Alice offers codecs `foo` and `bar`, and Bob answers with only codec `foo`, thereby rejecting `bar`. If Bob were to now create a subsequent offer, his remote description still contains both `foo` and `bar`, so he would technically be allowed to offer both codecs, even though `bar` was previously rejected.

Is this intentional? It seems to go against the spirit of the rule. Perhaps the rule is meant to be "only include codecs present in the current local and remote descriptions".

For context, see: https://www.ietf.org/mail-archive/web/rtcweb/current/msg15697.html

**Comment by juberti** [comment]
Yes, your understanding is correct. The set to be offered should only include the subset of codecs that were negotiated.

**Comment by ibc** [comment]
>  The set to be offered should only include the subset of codecs that were negotiated.

Please, let's specify the context of those already negotiated codecs. Is that per m=line? is that per "RTP session" (whatever that means nowadays)? is that per full SDP?

If Alice just offers an m=audio line in the initial SDP offer and Bob answers, aren't they allowed to re-offer later with an added m=video line?

**Comment by ibc** [comment]
Also, are all those rules take into account multi-stream/participants when in a SFU scenario?

Alice may join a SFU conference by offering VP8 and H264 for her sending m=video section, and the SFU may reply with just VP8. But later when a new non-VP8 capable peer join the conference, the SFU may want to generate a re-offer for Alice in behalf of Bob with a new m=video line (a=sendonly) having just H264 codec (since Bob). Why is this illegal?

**Comment by ibc** [comment]
Related issue open in Chromium project: https://bugs.chromium.org/p/webrtc/issues/detail?id=5697

**Comment by taylor-b** [comment]
> Is that per m=line? is that per "RTP session" (whatever that means nowadays)? is that per full SDP?

I assumed it meant per m= line, though this should definitely be made clear. If my assumption is correct, there would be no issue in your later comment about an SFU adding a new m= line.

**Comment by ibc** [comment]
Thanks @taylor-b, it makes sense. Hope it's clarified in the spec.

**Comment by juberti** [comment]
Eric and I discussed today and I think the simplest change is to reference "the codecs present in the most recent answer", since that is the set of negotiated codecs. 

This should also be extended to the other bullets in 5.2.2 regarding negotiated properties.

**Comment by ibc** [comment]
> Eric and I discussed today and I think the simplest change is to reference "the codecs present in the most recent answer", since that is the set of negotiated codecs.

@juberti @ekr have you consider this scenario?:

* Alice may join a SFU conference by offering VP8 and H264 in her m=video (a=sendrecv), and the SFU may reply with just VP8 to ensure VP8 is used.
* Later, when a new non-VP8 capable peer joins the conference, the SFU may want to generate a re-offer for Alice by changing the codec to just H264.

With your change this is not legal. And I don't agree because it is an artificial limitation.

If we play the SDP game we should follow its rules. Does some RFC state that my suggested use case is illegal?

**Comment by juberti** [comment]
SFUs can do whatever they want. This spec is for JSEP implementations.

**Comment by ibc** [comment]
@juberti please, re-read my concern. It's about a valid SDP received by WebRTC browser.

**Comment by juberti** [comment]
Nothing prohibits the SFU from re-offering with H.264 + VP8, and the JSEP endpoint will handle this properly when it is received (generating an appropriate answer). 

The text being discussed here only affects the offers that JSEP generates.

**Comment by fluffy** [comment]
Re-openened due to discution on ietf list

**Comment by rshpount** [comment]
I think re-offer should contain the full set of codecs or this will cause interesting problems with 3pcc or SIP interop. The order of codecs should ideally be the same as currently negotiated to avoid codec flips. Essentially, what this means all supported codecs not currently negotiated should be added to the end of the offer after the currently negotiated ones.

**Comment by cdh4u** [comment]
Note that there is a generic problem about what we mean by "codec". For example, assuming we are only allowed to re-offer previously negotiated codecs (e.g. H.264), are we still allowed to re-offer a different H.264 "codec configuration"?

**Comment by ibc** [comment]
Related: can an endpoint re-offer by introducing new PT values that were not present in previous O/A?

**Comment by fluffy** [comment]
So one way to create a re-offer with all the new PT value that got removed in prev O/A is to create a  new PC, add all the steams/tracks to it. Then call create offer to create a "fresh" offer and use that for the offer. However, might be a bit complicated with what happens with incoming media, some of it would be incoming on old PC then move to new PC.

**Comment by rshpount** [comment]
This is even more complicated since you are not supposed to re-use the old PT in the new offer for different codecs. Since this is a completely new offer with no information about previously used PT, this restriction is hard to enforce.

**Comment by ibc** [comment]
> So one way to create a re-offer with all the new PT value that got removed in prev O/A is to create a new PC, add all the steams/tracks to it

Honestly, I don't think such an approach is acceptable.

**Comment by fluffy** [comment]
Hmm - I see your point.

**Comment by ibc** [comment]
I don't want to renegotiate ICE+DTLS. I just want to add my audio/video on an already connected transport, and make it work :)

**Comment by taylor-b** [comment]
From [the IETF 97 minutes](https://tools.ietf.org/wg/rtcweb/minutes?item=minutes-97-rtcweb-00.html), it appears the decision was to keep the spec as-is (only negotiate a subset), and possibly add an API in the future for adding codecs.

Also, non-JSEP endpoints *are* still allowed to add codecs, which is covered by #409.

So, I think this issue can be closed.

**Comment by rshpount** [comment]
This will definitely break things.

**Comment by taylor-b** [comment]
I just listened to the recording, and the conclusion appeared to be "no one in the room objects to this, though some people on the mailing list might, so we'll continue discussion there". Did that discussion never happen? If so, it needs to. I'll reopen.

**Comment by rshpount** [comment]
I do not think discussion ever happened on the list. My issue is that there is no API to generate a new offer which reuses currently negotiated payload types but lists all available codecs.  Without this functionality third party call control is impossible to implement and it is impossible to simulate SIP INVITE without the body which will affect interop.

**Comment by pthatcherg** [comment]
A possible addition to the API to allow reoffer codecs that were excluded previously:

https://github.com/w3c/webrtc-pc/pull/988

And the JSEP part:

https://github.com/rtcweb-wg/jsep/pull/493

**Comment by taylor-b** [comment]
Assigning Peter since his PR seems like the way forward.

**Comment by juberti** [comment]
I do worry that handling this in an ad hoc fashion is going to bite us, e.g. what if in the future a SFU wants a new offer with all supported RTP header extensions? I'm going to go dig up the initial issue that led to the approach that is currently in the document and see if it helps frame the situation.

**Comment by juberti** [comment]
**TL; DR:** After researching this issue, I think we should resolve it by always putting any supported-but-not-previously-negotiated codecs into the offer, but *after* the previously-negotiated codecs. [I will note that this is what @rshpount advocated for in August.]

Full analysis
------------

OK, I trawled through the archives, and to the best of my understanding there are three issues:
1. How to handle the case where EP1 (offerer) supports codecs [A, B, C] (in that preference order) and EP2 (answerer) supports [D, C, B] (in that order). See https://github.com/rtcweb-wg/jsep/pull/269 for more info.
2. Interactions with the codec ordering API discussed in #424 and related - how to prevent problems when both sides don't agree on ordering.
3. Unnecessary SDP traffic (resending stuff that won't be used)

Turning our attention to 1), we can consider left with the situation where EP1 and EP2 negotiate [B, C], via default 3264 rules, and EP2 now wants to reoffer. 
- Under the current text, EP2 would have to reoffer [B, C], which prevents any change to the codec in use by either side, but doesn't allow for use of D.
- However, we definitely do not want to reoffer [D, C, B], since that will lead to [C, B] being negotiated, i.e., it will cause a codec switch.
- That leaves [D, B, C] or [B, C, D] as the only reasonable options to consider, i.e. "put new ones first" or "put new ones last".
- "New ones first" is complicated if we consider the case where EP2 suddenly supports [D, C, B, A]. Then the offer would become [D, A, B, C], and result in [A, B, C] being negotiated, which is probably not what EP2 wants.
- "New ones last" results in [B, C, D, A] being offered in the situation above, and [B, C, A] being negotiated. No codec switch occurs, unless explicit priority tweaking is performed.

Ergo, "new ones last" seems like a reasonable policy that should address 1).
2) I think can be handled by the solution I am about to propose in #424.
3) I believe to be of minor importance. In the common case (e.g. when communicating between like endpoints), there will be no changes, and even in interop cases, the difference should be insignificant.

Following up on my earlier comment, what about other things that might need to be renegotiated, such as a=rtcp-fb fields, and a=extmap extensions? It seems likely that if we are going to allow new codecs, we need to allow new a=rtcp-fb (which can be associated with a PT). And if we're going to do that, new a=extmap seems largely harmless. 

However, I think we should draw the line at any changes to TRANSPORT attributes, e.g. a=rtcp-rsize, etc. So the changes to S 5.2.2 would basically just be to incorporate the policy mentioned above into the first 3 bullets on page 45 (and any related changes needed in the subsequent-answer section).

The End.

**Comment by ekr** [comment]
I can live with this. I would note that in bullet 3 above you could also do "B D C" but I think we can agree that that would be silly.

**Comment by fluffy** [comment]
sounds good to me. Lots of SIP stuff does exactly that so I don't foresee any problem wiht his.

**Comment by rshpount** [comment]
I can leave with this as well. 

I just want to remind about two arguments that I got when I proposed this solution:
a. Allocating a new codec can consume resources. For instance some video codecs have hardware resources required for their operation which need to be reserved before codec is offered. If codec is offered but never used, this can lead to waste of resources.
b. In case of SIP, complete list of codecs is not produced for every offer. It is only produced for the offer which is generated in response to INVITE with no SDP body. Proposed behavior does not allow to exactly match currently implemented SIP end point behavior.

As I have stated, I am fine with the proposed solution. I just want to make sure these two arguments do not cause any issues for anybody else.

**Comment by juberti** [comment]
@fluffy, wdyt about the most recent comment from @rshpount ?

As I see it we can either do what I propose here always, or only upon a createOffer flag (as @pthatcherg suggested). Always seems simpler, if we aren't worried about the issues Roman raises.

**Comment by fluffy** [comment]
So what we sorted on editor call is we are going to make this happen by default for now. If a strong use case for having a knob to control it comes up, we can consider that. One point is you can use codec prefs to eliminate codecs so they don't show up in future offers.

**Comment by juberti** [comment]
@pthatcherg, can you update your PR to reflect this consensus?

**Comment by ibc** [comment]
Hi, sorry as I couldn't track all the conclusions in this issue. However I'd like to confirm whether the current Chrome behavior is correct (according to the conclusion above) or not:

* Chrome is offered with VP8 PT 123.
* Chrome answers with VP8 PT 123 (including a sending video track) and no other video codecs.
* Chrome removes the video track from its local MediaStream.
* Chrome re-adds the same video track into its local MediaStream.
* Chrome calls `createOffer()`.

The new offer contains VP8 PT 123, but also VP9 PT 101, H264 etc etc etc. I didn't expect it since the initial SDP O/A did not contain them. Of course, there is a single BUNDLED transport.

If this behavior correct and should Chrome fix it?

**Comment by fluffy** [comment]
I think that behavior is correct by the new definition which is roughly reoffers will offer everything the browser can do at that point of time, but will put the stuff currently in use first.

**Comment by ibc** [comment]
Fine with me :)

**Comment by juberti** [comment]
Note that removal/readdition from a MediaStream should have no impact on offers/answers. IOW, the case you describe above is essentially: offer with all codecs, answer with just VP8, reoffer contains all codecs.

**Comment by juberti** [comment]
See also #602 as a trivial followup on this issue.

**Comment by juberti** [comment]
Which has been merged, bringing this issue, finally, to a close.

**Comment by ibc** [comment]
> Note that removal/readdition from a MediaStream should have no impact on offers/answers

What about if the video track I add to the local stream is not the same as the one I previously removed from it? For example, it is a video track retrieved via `getUserMedia()` with different video resolution/constraints settings.

**Comment by juberti** [comment]
It has no impact. Only tracks given to addTrack or setTrack are used.

**Comment by ibc** [comment]
> It has no impact. Only tracks given to addTrack or setTrack are used.

I did not understand you, sorry. I know that, according to the new API, changes in the local `MediaStream` do not affect what the `PeerConnection` sends (although current Chrome does not behave like that and reacts on track addition/removal into the `MediaStream`).

Thanks.

**Comment by ibc** [comment]
Just to be 200% sure, I would like to confirm that the following behaviour in Firefox is wrong. The scenario is the same as [above](#issuecomment-279245497) but now with Firefox: 

* Firefox is offered with VP8 PT 123.
* Firefox answers with VP8 PT 123 (including a sending video track) and no other video codecs.
* Firefox removes the video track from its PeerConnection.
* Firefox adds another local video track into its PeerConnection.
* Firefox calls `createOffer()`.

The new offer contains VP8 PT 120 instead of 123 (wrong).

Bug/issue detailed here: https://bugzilla.mozilla.org/show_bug.cgi?id=1339234

**Comment by ekr** [comment]
What exact text do you believe prohibits this?

**Comment by rshpount** [comment]
Per RFC 3264 this is allowed. What is not allowed is to use PT 123 for anything except VP8.

Per [RFC3264 8.3.2 Changing the Set of Media Formats](https://tools.ietf.org/html/rfc3264#section-8.3.2):

> However, in the case of RTP, the mapping from a particular dynamic payload type number to a particular codec within that media stream MUST NOT change for the duration of a session.  For example, if A generates an offer with G.711 assigned to dynamic payload type number 46, payload type number 46 MUST refer to G.711 from that point forward in any offers or answers for that media stream within the session.  However, it is acceptable for multiple payload type numbers to be mapped to the same codec, so that an updated offer could also use payload type number 72 for G.711.

**Comment by ibc** [comment]
Well, probably nothing prohibits a re-offer to include new PT values pointing to same codec+configurations as previously negotiated. Said that, a re-offer cannot offer a previously used PT for other codec+configuration.

So IMHO this could be simplified by mandating reusing same PT values as before.

**Comment by ekr** [comment]
I'm not seeing a lot of value in requiring JSEP impls to behave this way.

**Comment by fluffy** [comment]
Just noting that if it was a different m line, then it could use any PT it wanted for VP8. And in the above example, it is fine for it to use 120 in the 2nd offer as long as 120 had not previously been used for something other than VP8. 

To answer EKR question, the only thing of value here is following use case. In offer/answer 1, we have PT 120 => H264 and  PT 123 => VP8. If in offer/answer 2 (on same m-lines, same mid, same  blah blah blah) we swapped to PT120 => VP8, if we got a late media packet that was sent before the switch, but arrived after the switch, it might get delivered to the VP8 codec when actually it was encoded with H.264. I think that is all the is meant to stop. And I have allwasy viewed that trying to sop this is silly because what is going to happen is that late packet is going to get discarded. If sending a 264 packet to your VP8 codecs causes anything to happen other than it getting discarded, well seem like that is a problem.

**Comment by juberti** [comment]
The behavior mentioned by @ibc does seem complicated, since the offerer needs to decode VP8 on both PT 120 and 123 at least until the O/A completes, but I agree it appears to be legal, and AFAICT it does not create any interoperability problems.

**Comment by ibc** [comment]
> Just noting that if it was a different m line, then it could use any PT it wanted for VP8

Uh no no. That's not possible. As long as such a m line belongs to the same BUNDLE group, all the m lines in that group are part of the same RTP session®️ [1] and, hence, their PT values must be unique (in the sense that they must point to the same codec+configuration®️).

I already asked about this exact subject several months ago. The given rationale was:

"Existing RTP intermediaries®️ [2] may rely on a single 1:1 mapping between PT and codec+configuration when carried over the same IP:port tuple, this is, the same RTP session"

Those are the related threads:

* https://www.ietf.org/mail-archive/web/mmusic/current/msg16017.html
* https://lists.w3.org/Archives/Public/public-webrtc/2016Mar/0021.html
* https://www.ietf.org/mail-archive/web/avt/current/msg17059.html


[1] Future generations may experience difficulties in understanding the concept of "RTP session".
[2] Although no one has ever seen them in a WebRTC environment, they may exist.

**Comment by ibc** [comment]
> The behavior mentioned by @ibc does seem complicated, since the offerer needs to decode VP8 on both PT 120 and 123 at least until the O/A completes, but I agree it appears to be legal, and AFAICT it does not create any interoperability problems.

The sad thing of all of this is that, at the end, for things to work the offerer and re-offerer must always be the same peer.

**Comment by fluffy** [comment]
I think we definitely need to be able to have the reoffer move the media to a different IP since we do that all the time on non webrtc stuff. I might be missing what the issue is here but I want that to work.

**Comment by ibc** [comment]
Move the m= section into a separate transport just because it includes PT values already present in the other m= sections but pointing to different codec+configurations?

**Comment by ekr** [comment]
I don't see how any of this is a an issue. The rule is you can't reuse the PT. There's no rule that you can't reuse the codec with a new PT.

**Comment by ibc** [comment]
> There's no rule that you can't reuse the codec with a new PT.

That's right, but it's also true that it causes errors in implementations (such as when Firefox renegotiates with Chrome by offering two different PT for Opus and Chrome assumes receipt of just the first PT).
